src/gui_common/CustomDialog.cs

```diff
@@ -0,0 +1,489 @@
+    /// </summary>
+    [Export]
+    public string WindowTitle
+    {
+        get => windowTitle;
+        set
+        {
+            windowTitle = value;
+            MinimumSizeChanged();
+            Update();
+        }
+    }
+
+    /// <summary>
+    ///   If true, the user can resize the window.
+    /// </summary>
+    [Export]
+    public bool Resizable { get; set; }
+
+    /// <summary>
+    ///   If true, the window's position is clamped inside the screen so it doesn't go out of bounds.
+    /// </summary>
+    [Export]
+    public bool BoundToScreenArea { get; set; } = true;
+
+    [Export]
+    public bool ExclusiveAllowCloseOnEscape { get; private set; } = true;
+
+    public override void _EnterTree()
+    {
+        // To make popup rect readjustment react to window resizing
+        if (!GetTree().Root.IsConnected("size_changed", this, nameof(OnViewportResized)))
+            GetTree().Root.Connect("size_changed", this, nameof(OnViewportResized));
+
+        SetupCloseButton();
+        UpdateChildRects();
+    }
+
+    public override void _Notification(int what)
+    {
+        switch (what)
+        {
+            case NotificationResized:
+            {
+                UpdateChildRects();
+                break;
+            }
+
+            case NotificationVisibilityChanged:
+            {
+                if (Visible)
+                {
+                    OnShown();
+                }
+                else
+                {
+                    OnHidden();
+                }
+
+                UpdateChildRects();
+                break;
+            }
+
+            case NotificationMouseExit:
+            {
+                // Reset the mouse cursor when leaving the resizable window border.
+                if (Resizable && dragType == DragType.None)
+                {
+                    if (MouseDefaultCursorShape != CursorShape.Arrow)
+                        MouseDefaultCursorShape = CursorShape.Arrow;
+                }
+
+                break;
+            }
+        }
+    }
+
+    public override void _Draw()
+    {
+        var panel = GetStylebox("custom_panel", "WindowDialog");
+        var titleBarPanel = GetStylebox("custom_titlebar", "WindowDialog");
+        var titleBarHeight = GetConstant("custom_titlebar_height", "WindowDialog");
+
+        // Draw background panels
+        DrawStyleBox(panel, new Rect2(
+            new Vector2(0, -titleBarHeight), new Vector2(RectSize.x, RectSize.y + titleBarHeight)));
+
+        DrawStyleBox(titleBarPanel, new Rect2(
+            new Vector2(3, -titleBarHeight + 3), new Vector2(RectSize.x - 6, titleBarHeight - 3)));
+
+        // Draw title in the title bar
+        var titleFont = GetFont("custom_title_font", "WindowDialog");
+        var titleHeight = GetConstant("custom_title_height", "WindowDialog");
+        var titleColor = GetColor("custom_title_color", "WindowDialog");
+
+        var translated = TranslationServer.Translate(WindowTitle);
```

Comment:
I think this should be cached when the window title is updated and not called on each draw.